### PR TITLE
GTK UI will now update when changes are detected to files.

### PR DESCRIPTION
**Pull request by sebastiansam55**
Prior to this a warning message would pop on screen when any changes not made by autokey were detected, now it detects and adds a watch on the new folder/files. It also updates the GtkTreeView with the new items (along with removing deleted or changing renamed items). 

Stores the current open file/folder in the global config file and uses it to reopen where you left off when reopening UI. Similarly stores the state of the GtkTree expanded/collapsed to restore it to the way that it was when the program was closed, also stored in the global config file.

### CODE REVIEW

File: lib/autokey/gtkapp.py [hunk 1]
```diff
@@ -171,12 +171,20 @@ def hotkey_removed(self, item):
     def path_created_or_modified(self, path):
         time.sleep(0.5)
         changed = self.configManager.path_created_or_modified(path)
+        if not self.monitor.has_watch(path) and os.path.isdir(path):
```
Comment (BlueDrink9) [inline lib/autokey/gtkapp.py:188]: These sections could probably be extracted as `add_open_file_watch` and `remove_open_file_watch` functions

File: lib/autokey/gtkui/configwindow.py [hunk 2]
```diff
@@ -868,18 +871,17 @@ def set_dirty(self, dirty):
...
-close and reopen the AutoKey window.\nThis message is only shown once per session.")
-            dlg = Gtk.MessageDialog(self.ui, type=Gtk.MessageType.QUESTION, buttons=Gtk.ButtonsType.OK,
-                                   message_format= _("Configuration has been changed on disk."))
-            dlg.format_secondary_text(msg)
-
-            dlg.run()
-            dlg.destroy()
-            Gdk.threads_leave()
-            self.__warnedOfChanges = True
+        #save tree view selection
```
Comment (BlueDrink9) [inline lib/autokey/gtkui/configwindow.py:898]: Probably worth putting a logging message in here. Probably at INFO level, eg `"Modifications detected to open files. Reloading..."`

File: lib/autokey/gtkui/configwindow.py [hunk 3]
```diff
@@ -1330,13 +1340,35 @@ def on_rename(self, widget, data=None):
...
+        """This function is called when a row is double clicked"""
         if widget.row_expanded(path):
             widget.collapse_row(path)
         else:
             widget.expand_row(path, False)
 
     def on_treeWidget_row_collapsed(self, widget, tIter, path, data=None):
         widget.columns_autosize()
+        p = path.to_string()
+        path_len = len(p)
```
Comment (BlueDrink9) [inline lib/autokey/gtkui/configwindow.py]: This needs to be a new function with a descriptive name. I can't tell what this is doing at a glance.

**Review by BlueDrink9 — CHANGES_REQUESTED** [review]

**Comment by BlueDrink9** [comment]
Such a great feature, can't wait to get this merged. Thank you!

**Comment by BlueDrink9** [comment]
Hey @sebastiansam55 how much time to work on this do you have atm? It would be cool to merge this before the beta release.

**Comment by sebastiansam55** [comment]
I'll try and get back up to speed on this, I'm afraid when I let things sit for this long its like starting from scratch because my memory is terrible

**Comment by josephj11** [comment]
Take a look at #533 and #534 to see if they affect this. (If you feel it's worth addressing at all.)

**Comment by BlueDrink9** [comment]
Sure thing. I don't think changes needed are massive. I would have done them myself but wasn't totally sure of the purpose of all of the code.

**Comment by BlueDrink9** [comment]
@sebastiansam55 can you please test out the change I made in 3bbcefa8c4801dbace0cb924b39ab83d405567a1, or at least confirm the change looks equivalent?

Once that's done I think we will be good to merge

**Comment by BlueDrink9** [comment]
(Also just to note that I rebased this on to develop and fixed the merge conflict, in case you are wondering why we are both authors for that recent change)

File: lib/autokey/gtkui/configwindow.py [hunk 4]
```diff
@@ -428,179 +428,6 @@ def set_dirty(self):
...
-
-    def redo(self):
-        self.buffer.redo()
-        self.parentWindow.set_undo_available(self.buffer.can_undo())
-        self.parentWindow.set_redo_available(self.buffer.can_redo())
-
-    def on_modified(self, widget, data=None):
-        self.set_dirty()
-        self.parentWindow.set_undo_available(self.buffer.can_undo())
-        self.parentWindow.set_redo_available(self.buffer.can_redo())
```
Comment (sebastiansam55) [inline lib/autokey/gtkui/configwindow.py]: Did you mean to remove all of this code? It (based on my understanding) is still necessary, this does look like something that could be abstracted but that hasn't happened in this branch yet from what I can tell...

@BlueDrink9

Comment (BlueDrink9) [reply lib/autokey/gtkui/configwindow.py]: Oh dang, thank you. This didn't even look changed to me, and definitely shouldn't have been removed

File: lib/autokey/gtkui/configwindow.py [hunk 5]
```diff
@@ -1360,13 +1371,36 @@ def on_rename(self, widget, data=None):
...
 
     def on_treeWidget_row_collapsed(self, widget, tIter, path, data=None):
         widget.columns_autosize()
+        p = path.to_string()
+        if len(p) == 1: #closing one of the base dirs
+            self.__hide_row_with_path_up_to(p, 1)
+            return
+        self.__hide_row_with_path_up_to(p, len(p))
+
+    def __hide_row_with_path_up_to(self, pathStr, up_to):
```
Comment (BlueDrink9) [inline lib/autokey/gtkui/configwindow.py:1388]: @sebastiansam55 does this seem like a reasonable function name? I wasn't totally clear on what this code actually as meant to do (differences between trees, rows, and paths etc).

Comment (sebastiansam55) [reply lib/autokey/gtkui/configwindow.py:1388]: a bit verbose but does the job haha

Comment (BlueDrink9) [reply lib/autokey/gtkui/configwindow.py:1388]: Better explicit than implicit. Imo so long as we aren't at Java or Apple C levels, more verbosity is good

**Comment by BlueDrink9** [comment]
@sebastiansam55 Two bugs to report with this feature on 9fd9107

1. Changing a phrase once seems to trigger 3 reloads before the UI updates. This is quite slow.
2. Making many changes in quick succession causes a **segfault**. Not much of a useful python stack trace either. See if you can reproduce this yourself: click the path for a phrase and open it in an editor, and quickly type one char, save, and repeat 2-3 times fast. I think it might be to do with threading. Maybe we explicitly unwatch a file when reloading, and only watch it again when reloading completes?

**Pull request merged**